Database architecture in SQL Server #2 (Fixes #2)

SaidAbdullahAbid opened the pull request:
**Added the following tables to the database:**

* `DataValuesAudit`
* `DataValues`
* `IndicatorChartTypes`
* `ChartConfigs`
* `Indicators`
* `Locations`
* `Calendars`
* `Unites`

**Included files:**

1. **`table_script.sql`** – Contains the SQL script to create the new tables.
2. **`triggers_script.sql`** – Contains two triggers:

   * `trg_CalculateGrowthAndTotalMetrics` on the `DataValues` table: Calculates the `growth` and `total` columns on `INSERT`, `UPDATE`, or `DELETE` based on `ChartConfigs` columns `CalculateGrowthBy` and `CalculateTotalBy`. If equal, growth is calculated based on `legend` or `indicator` in percentage; total is considered the same.
   * `trg_RecalculateMetricsOnConfigChange` on the `ChartConfigs` table: When `CalculatedByGrowth` changes (indicator or legend), it recalculates growth for affected rows in `DataValues` and updates the corresponding columns.
3. **`seeding_queries_scripts.sql`** – Contains real data based on the attached chart images.
4. **`testing_queries_scripts.sql`** – Contains queries to test whether the triggers are working correctly.
5. **`PKU_dashboard_database_scripts.sql`** – Complete database schema with default data.

**Important:**

* If you already have any of these tables in your database, please modify or remove them before running the above scripts to avoid conflicts.
![First_Image](https://github.com/user-attachments/assets/84749dc7-00b5-49ee-85a9-29533d245244)
![second_image](https://github.com/user-attachments/assets/ecc31465-50f0-402a-aed7-2442058a10fe)

sayedkhalid-sultani commented:
> @sayedkhalidsultani Dear Sultani Saib, I have added another table for the source with a foreign key to the indicator. I have also added the Longitude and Latitude properties to the Location table. If needed, we can create another table to establish a many-to-many relationship between indicators and sources. Could you please review it and also specify what kind of modification is needed for the Calendar table?

@SaidAbdullahAbid  Dear abid seab here are some suggestion.
**GIS Side** 
for GIS here are some samples , 
Two Indicator Map tab from world bank will provide you some idea.
https://www.worldbank.org/en/data/interactive/2019/08/01/afghanistan-interactive-province-level-visualization
here is another example
https://response.reliefweb.int/afghanistan/natural-disasters-dashboard
if you see these two examples on first layer the map is choropleth map and second it is bubble chart, on unocha dashboard(second one ) the bubble size is based on amount the bigger the figure the bigger the bubble ( i need a single bubble in center of provinces not multiple as provided here) on the first dashboard (world bank one) the bubble colors are set by amount. you can check for both which one looks good so you can plot population (first indicator) against poverty (second ex) across 34 provinces.
**Calender Table** 
Calendar table add below columns , so we could specify on x , axis of the chart the format we want to view the chart if the type is date type , also fill your calendar table for date between 2020-2030 for time being .

-- Month label /Short ex Jan
-- Quarter Label/Short  ex Q1
-- YearQuarterlabel Q1/2000 
-- Month/Year Label  Jan/2000
you will see the usage of these across some samples which i attach.
_for charts please test database structure against some samples i am sharing here at least test it against 10 different chart types ._
The charts contains multiple series (bar and line together) , multi level x axis on barchart , multiple Y series (right and left) , multiple sources on linechart an many others please review them and test database against these charts.
<img width="446" height="236" alt="{0D3CDAA7-590A-4EC3-BBD2-084F928709DE}" src="https://github.com/user-attachments/assets/83542de2-96be-46fd-acfc-95f1353b2de9" />
<img width="474" height="205" alt="{3C44DE3F-23C1-404F-98CC-6312471A6BDD}" src="https://github.com/user-attachments/assets/0467e341-9123-404c-9d58-dabc6695df3f" />
<img width="223" height="188" alt="{7D436C7B-E282-4195-8F01-88DD7554FB23}" src="https://github.com/user-attachments/assets/762cf597-6fd4-490f-8ba9-8860834c114a" />
<img width="408" height="236" alt="{80221DEE-1A61-49F6-966A-4D8BE10A85F6}" src="https://github.com/user-attachments/assets/de28a5e1-99cb-4691-9fd3-27eb9703982c" />
<img width="225" height="225" alt="{B885164C-443A-45DB-86AD-614011C9A2FC}" src="https://github.com/user-attachments/assets/dd34367e-6cde-4f0c-aef1-234871a06d21" />
<img width="231" height="151" alt="{C60CB869-AB64-41A8-ADB2-9CE06FC374DF}" src="https://github.com/user-attachments/assets/394c3ed1-c7f7-4735-a2fa-db14288b5715" />
<img width="296" height="121" alt="{D07A1932-54C9-4690-9B82-0C64FA663308}" src="https://github.com/user-attachments/assets/35973d2e-d59a-4875-81b5-e805b87b9484" />
<img width="223" height="198" alt="{E176968F-9BE0-41E3-8CB9-6737B26B995D}" src="https://github.com/user-attachments/assets/39df95e2-e442-4ef5-9e09-e3b20ac3ea1c" />
<img width="410" height="235" alt="{F4C829BC-B21B-4605-BFD0-9338BAA8C081}" src="https://github.com/user-attachments/assets/706a0654-09b2-4485-b72b-2863e01f159f" />
<img width="301" height="199" alt="{F732E5C3-129C-456B-815B-97DACA44AFCE}" src="https://github.com/user-attachments/assets/1feb0191-689e-4d4f-a6c5-8de57e946951" />
<img width="479" height="281" alt="{F44720E6-B6CB-46DC-B2C9-8769F3891FF4}" src="https://github.com/user-attachments/assets/d9ce8916-b060-4709-9dd5-2972f875bd75" />
the values should also store below columns atleast 
1. % Growth /Changes  since last period (ex now i am in aug 2025 so previous period is last available data based on date say Mar 2025)
2. % Growth/Changes since last Year same period these will be required for currency exchange rate tracking and the rest. (ex currently i am in dec 2025 , i have data on dec 2024 so  your trigger should calculate % difference and updated table column.
3. % of Total across items with same Parent  as ex we have kabul province , and district kabul, musayee , paghman with population so for districts kabul ,musayee i need to see how much in terms of % of total the population based on kabul population and the rest makes say kabul is 50% , 30% paghman , 10% the rest . 
As these values are stored in database it will be easily displayed in popups when we move cursor across bars and line chart on the front end.
Hope i am clear with the requirements if anything is not clear do let me know .

SaidAbdullahAbid commented:
> > @sayedkhalidsultani Dear Sultani Saib, I have added another table for the source with a foreign key to the indicator. I have also added the Longitude and Latitude properties to the Location table. If needed, we can create another table to establish a many-to-many relationship between indicators and sources. Could you please review it and also specify what kind of modification is needed for the Calendar table?
> 
> @SaidAbdullahAbid Dear abid seab here are some suggestion. **GIS Side** for GIS here are some samples , Two Indicator Map tab from world bank will provide you some idea. https://www.worldbank.org/en/data/interactive/2019/08/01/afghanistan-interactive-province-level-visualization here is another example https://response.reliefweb.int/afghanistan/natural-disasters-dashboard if you see these two examples on first layer the map is choropleth map and second it is bubble chart, on unocha dashboard(second one ) the bubble size is based on amount the bigger the figure the bigger the bubble ( i need a single bubble in center of provinces not multiple as provided here) on the first dashboard (world bank one) the bubble colors are set by amount. you can check for both which one looks good so you can plot population (first indicator) against poverty (second ex) across 34 provinces.
> 
> _We do not need to store gps points in our database it will be retrieved from aims system._
> 
> **Calender Table** Calendar table add below columns , so we could specify on x , axis of the chart the format we want to view the chart if the type is date type , also fill your calendar table for date between 2020-2030 for time being .
> 
> -- Month label /Short ex Jan -- Quarter Label/Short ex Q1 -- YearQuarterlabel Q1/2000 -- Month/Year Label Jan/2000 you will see the usage of these across some samples which i attach. _for charts please test database structure against some samples i am sharing here at least test it against 10 different chart types ._ The charts contains multiple series (bar and line together) , multi level x axis on barchart , multiple Y series (right and left) , multiple sources on linechart an many others please review them and test database against these charts. <img alt="{0D3CDAA7-590A-4EC3-BBD2-084F928709DE}" width="446" height="236" src="https://private-user-images.githubusercontent.com/225797809/485970170-83542de2-96be-46fd-acfc-95f1353b2de9.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTczMTUyMjUsIm5iZiI6MTc1NzMxNDkyNSwicGF0aCI6Ii8yMjU3OTc4MDkvNDg1OTcwMTcwLTgzNTQyZGUyLTk2YmUtNDZmZC1hY2ZjLTk1ZjEzNTNiMmRlOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkwOFQwNzAyMDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yY2RhOTQ1ZGUxY2VkMmQ5ZTI1ZDdmNTY5MTcwN2M5MjVhODk4MmEwZDE5ZTUzMjdmNmExODVkOWM4YzA2OWIwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.pERtiq2INMyWqifBjGpfviABAWvieIZ-p-pyKka1CDY"> <img alt="{3C44DE3F-23C1-404F-98CC-6312471A6BDD}" width="474" height="205" src="https://private-user-images.githubusercontent.com/225797809/485970184-0467e341-9123-404c-9d58-dabc6695df3f.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTczMTUyMjUsIm5iZiI6MTc1NzMxNDkyNSwicGF0aCI6Ii8yMjU3OTc4MDkvNDg1OTcwMTg0LTA0NjdlMzQxLTkxMjMtNDA0Yy05ZDU4LWRhYmM2Njk1ZGYzZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkwOFQwNzAyMDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xYzAwOGYwZjdlOGMzNWZiN2I1MWI0YzlmMThmODJjYmMwODQzNDc3MDY0NzAwNjhjMjUwN2I2NGYzNjQyYTlkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.9-YMYThk2ZzvwT0WXX47WjYGqYiB10h_KELTE32mTac"> <img alt="{7D436C7B-E282-4195-8F01-88DD7554FB23}" width="223" height="188" src="https://private-user-images.githubusercontent.com/225797809/485970191-762cf597-6fd4-490f-8ba9-8860834c114a.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTczMTUyMjUsIm5iZiI6MTc1NzMxNDkyNSwicGF0aCI6Ii8yMjU3OTc4MDkvNDg1OTcwMTkxLTc2MmNmNTk3LTZmZDQtNDkwZi04YmE5LTg4NjA4MzRjMTE0YS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkwOFQwNzAyMDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03MmU3Y2EzOWNjOTdiYTM0MjAwMWZjYmY0Y2QwNTFkY2YxMTk3NDVkMWZkMTdiOWEyYzM0MDY5NDBkZGMwOTQ3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.GXHgs99CsF4RoaNeE5VI2LozFOL9g5X7gAF-OfHmXr8"> <img alt="{80221DEE-1A61-49F6-966A-4D8BE10A85F6}" width="408" height="236" src="https://private-user-images.githubusercontent.com/225797809/485970193-de28a5e1-99cb-4691-9fd3-27eb9703982c.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTczMTUyMjUsIm5iZiI6MTc1NzMxNDkyNSwicGF0aCI6Ii8yMjU3OTc4MDkvNDg1OTcwMTkzLWRlMjhhNWUxLTk5Y2ItNDY5MS05ZmQzLTI3ZWI5NzAzOTgyYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkwOFQwNzAyMDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hNjZkMDdjZDZiNmQ1NTM0MTlmMjhmNjE4ZTY1ZjEwMGJhYTg3ZWViOWQ2NTUzMTE5OWQxZDNhMTA2ZmJmZDQ0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.qhRG_gSK6XAU5K95PKvIte2VqQy98C5Rtp1ij-DN2HQ"> <img alt="{B885164C-443A-45DB-86AD-614011C9A2FC}" width="225" height="225" src="https://private-user-images.githubusercontent.com/225797809/485970195-dd34367e-6cde-4f0c-aef1-234871a06d21.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTczMTUyMjUsIm5iZiI6MTc1NzMxNDkyNSwicGF0aCI6Ii8yMjU3OTc4MDkvNDg1OTcwMTk1LWRkMzQzNjdlLTZjZGUtNGYwYy1hZWYxLTIzNDg3MWEwNmQyMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkwOFQwNzAyMDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05ZTI3MTgwZWYwMDkyNGFlNWM1NzUzYTlkZjMxYmFiZGQxNGJmNWQxNmY2NzQyNTIxNmFmODQ4YmY4YTZiMWViJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.5i_gpBX-6Bf06jTzQRYEia7SV5rV42Y6Qb1BlgguCQY"> <img alt="{C60CB869-AB64-41A8-ADB2-9CE06FC374DF}" width="231" height="151" src="https://private-user-images.githubusercontent.com/225797809/485970197-394c3ed1-c7f7-4735-a2fa-db14288b5715.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTczMTUyMjUsIm5iZiI6MTc1NzMxNDkyNSwicGF0aCI6Ii8yMjU3OTc4MDkvNDg1OTcwMTk3LTM5NGMzZWQxLWM3ZjctNDczNS1hMmZhLWRiMTQyODhiNTcxNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkwOFQwNzAyMDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02YzY1MThkNWUwMGU0YWJkZjk1Yjg4ZjU1NDUwZWZhZDUxODY3OGVkYmMyM2EzMGY0NGMyZTI1MGEwM2EyZTQ2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.TrDCioj6O8Zr_4m5lLQCVGrBvD4cNzO7I_R7kwEr-ZA"> <img alt="{D07A1932-54C9-4690-9B82-0C64FA663308}" width="296" height="121" src="https://private-user-images.githubusercontent.com/225797809/485970203-35973d2e-d59a-4875-81b5-e805b87b9484.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTczMTUyMjUsIm5iZiI6MTc1NzMxNDkyNSwicGF0aCI6Ii8yMjU3OTc4MDkvNDg1OTcwMjAzLTM1OTczZDJlLWQ1OWEtNDg3NS04MWI1LWU4MDViODdiOTQ4NC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkwOFQwNzAyMDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wMDg2ZjcxNTFmYzM4Yzg2MDM5ZWM1N2JmZTJhMTc4Y2QwNzRmNGQ5MTAxOTE2N2Q0ZmRmZGI0NjQ2ZWM5YTU5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.DAi76F4mcAMa-_s3l3XHXuk-DuWrGRHC6Xpzll5J2sw"> <img alt="{E176968F-9BE0-41E3-8CB9-6737B26B995D}" width="223" height="198" src="https://private-user-images.githubusercontent.com/225797809/485970204-39df95e2-e442-4ef5-9e09-e3b20ac3ea1c.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTczMTUyMjUsIm5iZiI6MTc1NzMxNDkyNSwicGF0aCI6Ii8yMjU3OTc4MDkvNDg1OTcwMjA0LTM5ZGY5NWUyLWU0NDItNGVmNS05ZTA5LWUzYjIwYWMzZWExYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkwOFQwNzAyMDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01ZTY1NjUxMzViOWY4NDlmNTE1NDNkMDYwMjllNzk1ZDNiNzYwYjM1OTBhMGMyMzM3MjIwYjlmYjdhZjU0YjA2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Fq5Ax-GfFrzXV3zDvFjRix_toUBkK1E9dfBFNoEgkh0"> <img alt="{F4C829BC-B21B-4605-BFD0-9338BAA8C081}" width="410" height="235" src="https://private-user-images.githubusercontent.com/225797809/485970208-706a0654-09b2-4485-b72b-2863e01f159f.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTczMTUyMjUsIm5iZiI6MTc1NzMxNDkyNSwicGF0aCI6Ii8yMjU3OTc4MDkvNDg1OTcwMjA4LTcwNmEwNjU0LTA5YjItNDQ4NS1iNzJiLTI4NjNlMDFmMTU5Zi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkwOFQwNzAyMDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jMzA4ODA3MjkzMGNmNGM2MTQwYjBiYmY5ZmIyNThiYmUwOWQ5NGQyNzM2YjhmN2VjMGU5NzI4OTJlZjI3Y2VkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.w5LYwG5A2UgwAVijejFskKqUxETLu5652Ul1Jml-tKk"> <img alt="{F732E5C3-129C-456B-815B-97DACA44AFCE}" width="301" height="199" src="https://private-user-images.githubusercontent.com/225797809/485970211-1feb0191-689e-4d4f-a6c5-8de57e946951.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTczMTUyMjUsIm5iZiI6MTc1NzMxNDkyNSwicGF0aCI6Ii8yMjU3OTc4MDkvNDg1OTcwMjExLTFmZWIwMTkxLTY4OWUtNGQ0Zi1hNmM1LThkZTU3ZTk0Njk1MS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkwOFQwNzAyMDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yZThhZjkyMGY3MmY3ZWRhM2JmNWRjZGRmMjJkZTQ5ZDQ1NmIyZDhiMTg3ZmY1Zjg1ODZkNDAzOWFiYmUzYTFhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.0tpYXg79b37bdhYxHRP-ccaxJyRNlwwFd4ot9ZsNvjU"> <img alt="{F44720E6-B6CB-46DC-B2C9-8769F3891FF4}" width="479" height="281" src="https://private-user-images.githubusercontent.com/225797809/485970215-d9ce8916-b060-4709-9dd5-2972f875bd75.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTczMTUyMjUsIm5iZiI6MTc1NzMxNDkyNSwicGF0aCI6Ii8yMjU3OTc4MDkvNDg1OTcwMjE1LWQ5Y2U4OTE2LWIwNjAtNDcwOS05ZGQ1LTI5NzJmODc1YmQ3NS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkwOFQwNzAyMDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lZDQwZTVjNGUwODBmNTA2NWM4NDllNGYyMGQ4YWI1NzcxMGZkNGQzOTFhMWQ0NDYzY2VlNDJkYmJjMTc2MjdhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.isP_qGK_GurCjB8ZvA7r7p3g3iv1lf3dIaFUH2i-SFk"> the values should also store below columns atleast
> 
> 1. % Growth /Changes  since last period (ex now i am in aug 2025 so previous period is last available data based on date say Mar 2025)
> 2. % Growth/Changes since last Year same period these will be required for currency exchange rate tracking and the rest. (ex currently i am in dec 2025 , i have data on dec 2024 so  your trigger should calculate % difference and updated table column.
> 3. % of Total across items with same Parent  as ex we have kabul province , and district kabul, musayee , paghman with population so for districts kabul ,musayee i need to see how much in terms of % of total the population based on kabul population and the rest makes say kabul is 50% , 30% paghman , 10% the rest .
>    As these values are stored in database it will be easily displayed in popups when we move cursor across bars and line chart on the front end.
>    Hope i am clear with the requirements if anything is not clear do let me know .

For the GIS point, I removed that from the location table.
For the calendar table, I added the suggested columns (2020–2030).
For growth and total portion calculations, I updated the trigger. The previous trigger already handled this with some flexibility—based on `chartConfig` (e.g., if set to legend, it calculates at legend level; if set to indicator, it calculates at indicator level).
The database now looks normalized and redundant properties are removed.

Additionally, I restructured the T-SQL files as follows:

* **01\_Table.sql** → table structure definitions
* **02\_Stored\_Procedures.sql** → stored procedure definitions
* **03\_Triggers.sql** → trigger definitions
* **04\_Seeding.sql** → database seeding scripts
* **05\_Schema\_Modification.sql** → schema modification scripts

Let me know if you have any further suggestions.

SaidAbdullahAbid commented:
**Greeting Dear @sayedkhalid-sultani  Sahib,**

The **DB section** is now working. As per your request, I have added a **trigger** for inserting and updating values in the `DataTables` section.

* When a new value is inserted, the trigger:

  * Finds the previous values to calculate **growth since the last period**.
  * Calculates **growth compared to the same period last year**.
  * Computes the **total share (proportion) it holds within the same indicator**.

For each of these sections, I have also added separate **SQL functions** to handle the calculations more clearly and maintainably.

I tested it with some of the graphs you provided, and it is working perfectly. We will continue to make changes according to your instructions and requirements in the next steps.
